### DISCUSSION TITLE
Tutorial / Guide?

### DISCUSSION THREAD

**Discussion by ModischFabrications** [AssetRipper/AssetRipper#732] [General]
### Describe the new feature or enhancement

I stumbled into this tool and the community surrounding asset ripping and I feel a bit lost, even with extensive Unity knowledge. 
I couldn't find any explanation on how the tool is used, what it can and can't do and how common use cases can be dealt with. 

I think this repo could really benefit from a simple guide or walkthrough to ease the learning curve a bit. Feel free to discard noobs without any knowledge of gamedev, but I feel that the barrier to entry is exceptionally high at the moment. 

These are my use cases, which probably mirror other users wishes somewhat:
1. I saw a poster/picture/image/... while playing the game and want to extract it -> Texture2D to png
2. I want to 3D-print a model from the game -> Mesh, maybe even textures (A & NM)
3. (Someone) wants to make mods
4. Finding glitches/exploits in scripts might also be interesting. Adjacent is fact checking for wikis, e.g. drop chance of a certain item. -> reading scripts

My rough suggestions:

-------

## Features
...

## Installation 
Download latest release for your platform from releases page. Copy somewhere and pin/shortcut *.exe. 

## Usage
Find the *.exe launcher file of your game. Check that there is a "UnityPlayer.dll" in the root folder, which shows you that it's a unity game.

Open AssetRipper, make basic settings. Disable script import if you don't need it, because this will increase load times by X. 
Drag *.exe into the window and wait for at least 15min, checking the console window for errors. This should result in a hierarchy similar to this [image here]. 

Optional: Download detected unity version via context menu to reimport as a project. 

// Copy over directory structure explanation from https://github.com/imadr/Unity-game-hacking

// What are all these option dropdowns? Might want to nest and sort them. "no script export" shouldn't show me the script decompile type.

### Extract image
Import project without scripts (and audio, ...?). 

Fold out "resources.assets\Texture2D", navigate through entries. You can see a preview on the right. Select and press "Export\Export selected" to a folder of your choosing. 
// *Additional feature request, I might post a separate issue later: "Rightclick > Save To ..." would be a great addition here, i don't care about the exportedProject around it*

Use a search tool, e.g. voidtools "everything", to find specific images, larger games usually have too many to go through manually. *art*, *poster*, ..., you get the idea. 

### Extract 3D printable model/mesh
1. Get Mesh?
2. Convert mesh to FBX?
3. Convert FBX to STL?

### Known issues
Encrypted files?

IL2Cpp code not loading?

RAM at max?

...?

**Comment by ds5678** [comment]
Moved because I could not see a clear condition for closure. 

> Rightclick > Save To

Ignoring UI stuff, this is basically #504 

> Tutorial for getting an STL file

I have marked this down in my Trello board for the YouTube channel.

> Download latest release for your platform from releases page. Copy somewhere and pin/shortcut *.exe.

An installer feels very Windows-centric, but it is somewhat warranted. For the current release of AssetRipper, 89% of downloads are for Windows. I will take this into consideration.

> search tool

This is far more complicated than most users expect.

> (Someone) wants to make mods

If you're referring to asset editing, I have no plans for that anytime soon (even after the hiatus).

> Encrypted files?

Out of scope for AssetRipper and copyright infringement

> IL2Cpp code not loading?

Most likely an upstream issue: https://github.com/SamboyCoding/Cpp2IL/issues

> RAM at max?

Not likely to change except perhaps for #504

**Reply by ModischFabrications** [reply]
>moved to discussion, closure

Suits me, was thinking about starting a discussion as well, but was unsure how you wanted to handle it. My initial take for a closure condition would have been to start documenting in the readme or webpage, but feel free to handle how it's easiest for you. 

> installer

I don't want to enforce documenting only the installer, but I would've liked some hint on where to look. Some use links on the website, some want you to clone the repo, some link executables in the releases page. Your metrics are understandable for me, because both Unity Development and Gaming is focused heavily on Windows. 

> search too complicated

It's definitely not something to recommend. It shows that a search within AssetRipper, prior to exporting, is even more useful imho. 

> out of scope, copyright or no plans, RAM

Just to clarify: I didn't want to request these features, but would suggest showing your hints, decisions and limitations somewhere. There will always be something that's out of scope, it's just helpful be stopped before investing a lot of time into a wrong solution. This might also be a good place for a comparison/link collection to other solutions that do fit the bill.

**Comment by ModischFabrications** [comment]
Additional hints I found:
- browse Texture2D's or GameObjects in sidebar to match unity level files to played surroundings
- loading level* as a file will pull in other files as needed
- Mesh Export Format -> GLB bypasses the need for unity, great for ripping assets. 
  - "Assets/GameObject/" of export should contain a fully textured mesh. 
    - Watch out: This might not be the best version of the mesh! "Assets/Mesh/LOD0_*" were the high quality ones in my case
  - "Assets/Scenes/..." will sometimes contain a fully textured level, but might be too large to open

**Comment by drdarkman-netizen** [comment]
For it to work you need to extract il2cpp if you have it, you need to organize the structure of the game if it was edited after the compilation so that assetripper organizes and extracts everything properly, place the .dl of il2cpp inside the structure, assetripper will identify the version of unity in the export tab you will be able to see it, once the engine is installed import the project select only 'ExportedProject' when it asks for errors simply ignore, this will create and open the project then fix the errors manually 'remember to distinguish between the version compatible for il2cpp and the normal version
